### DISCUSSION TITLE
fatal error: parser.h: No such file or directory

### DISCUSSION THREAD

**Discussion by flucrezia** [strongswan/strongswan#2438] [Q&A] [answered]
When compiling strongswan on multiple target platform using Docker containers to isolate builds (ubuntu 22.02, ubuntu 20.04, debian 11.6) we got a non-deterministic, subtle error, which appears only on builds triggered by a CI system similar to Github actions, but not in cloud: Gitea actions. We cannot reproduce on our laptops the failure and, as I said, on automated builds the failures are random. We believe this appears in case of high workloads, a lot of context switches on the host machine where the compilations run, possibly because of a race condition situation...
We try to build with **make -j1** without success. We use cmake to include strongswan on a broader project. 
Any help or hint on this would be much much appreciated ! Many thanks..

libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I/workspace/MyProject/modules/ipsec/external/strongswan-5.9.13/src/starter -I../.. -I/workspace/MyProject/modules/ipsec/external/strongswan-5.9.13/src/include -I/workspace/MyProject/modules/ipsec/external/strongswan-5.9.13/src/libstrongswan -I/workspace/MyProject/modules/ipsec/external/strongswan-5.9.13/src/libcharon -I/workspace/MyProject/modules/ipsec/external/strongswan-5.9.13/src/starter -I/workspace/MyProject/modules/ipsec/external/strongswan-5.9.13/src/stroke -DIPSEC_DIR=\"/usr/myproject/libexec/ipsec\" -DIPSEC_CONFDIR=\"/usr/myproject/etc\" -DIPSEC_PIDDIR=\"/usr/myproject/var/run\" -DIPSEC_EAPDIR=\"\" -DIPSEC_SCRIPT=\"ipsec\" -DDEV_RANDOM=\"/dev/random\" -DDEV_URANDOM=\"/dev/urandom\" -DDEBUG -DLOAD_WARNING -Wno-format -Wno-format-security -Wno-implicit-fallthrough -Wno-missing-field-initializers -Wno-pointer-sign -Wno-sign-compare -Wno-type-limits -Wno-unused-parameter -g -O2 -include /workspace/MyProject/build/modules/ipsec/external/strongswan-prefix/src/strongswan-build/config.h -MT parser/parser.lo -MD -MP -MF parser/.deps/parser.Tpo -c /workspace/MyProject/modules/ipsec/external/strongswan-5.9.13/src/starter/parser/parser.c  -fPIC -DPIC -o parser/.libs/parser.o
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I/workspace/MyProject/modules/ipsec/external/strongswan-5.9.13/src/starter -I../.. -I/workspace/MyProject/modules/ipsec/external/strongswan-5.9.13/src/include -I/workspace/MyProject/modules/ipsec/external/strongswan-5.9.13/src/libstrongswan -I/workspace/MyProject/modules/ipsec/external/strongswan-5.9.13/src/libcharon -I/workspace/MyProject/modules/ipsec/external/strongswan-5.9.13/src/starter -I/workspace/MyProject/modules/ipsec/external/strongswan-5.9.13/src/stroke -DIPSEC_DIR=\"/usr/myproject/libexec/ipsec\" -DIPSEC_CONFDIR=\"/usr/myproject/etc\" -DIPSEC_PIDDIR=\"/usr/myproject/var/run\" -DIPSEC_EAPDIR=\"\" -DIPSEC_SCRIPT=\"ipsec\" -DDEV_RANDOM=\"/dev/random\" -DDEV_URANDOM=\"/dev/urandom\" -DDEBUG -DLOAD_WARNING -Wno-format -Wno-format-security -Wno-implicit-fallthrough -Wno-missing-field-initializers -Wno-pointer-sign -Wno-sign-compare -Wno-type-limits -Wno-unused-parameter -g -O2 -include /workspace/MyProject/build/modules/ipsec/external/strongswan-prefix/src/strongswan-build/config.h -MT parser/lexer.lo -MD -MP -MF parser/.deps/lexer.Tpo -c parser/lexer.c  -fPIC -DPIC -o parser/.libs/lexer.o
**-- stderr output is:
/workspace/MyProject/modules/ipsec/external/strongswan-5.9.13/src/starter/parser/lexer.l:19:10: fatal error: parser.h: No such file or directory
   19 | #include "parser.h"
      |          ^~~~~~~~~~**
compilation terminated.
make[8]: *** [Makefile:704: parser/lexer.lo] Error 1
make[7]: *** [Makefile:731: all-recursive] Error 1
make[6]: *** [Makefile:546: all] Error 2
make[5]: *** [Makefile:542: all-recursive] Error 1
make[4]: *** [Makefile:604: all-recursive] Error 1
make[3]: *** [Makefile:515: all] Error 2
CMake Error at /workspace/MyProject/build/modules/ipsec/external/strongswan-prefix/src/strongswan-stamp/strongswan-build-Debug.cmake:47 (message):
  Stopping after outputting logs.
make[2]: *** [modules/ipsec/external/CMakeFiles/strongswan.dir/build.make:130: modules/ipsec/external/strongswan-prefix/src/strongswan-stamp/strongswan-build] Error 1
make[1]: *** [CMakeFiles/Makefile2:10090: modules/ipsec/external/CMakeFiles/strongswan.dir/all] Error 2
make: *** [Makefile:182: all] Error 2

**Comment by tobiasbrunner** [comment]
Can't really help you there. As you can see in the log, `parser.c` is compiled successfully. But since that file is generated together with `parser.h` by `bison`, I've no idea why the latter wouldn't exist when `lexer.c` is compiled afterwards.

**Reply by flucrezia** [reply, marked as answer]
Thank you for the quick reply, it triggered in me a solution (at least for now zero failures after tens of test runs) by viewing things in a different perspective :)
It seems the problem was in the CMakeLists.txt used to include strongswan.
I post here in case it will be useful for someone using cmake and remote automated builds.
Before it was:
```
ExternalProject_Add(strongswan
                    SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/strongswan-5.9.13 <--- Pay attention to this !
                    INSTALL_DIR ${CMAKE_BINARY_DIR}/host-install
                    LOG_CONFIGURE YES
                    LOG_BUILD YES
                    LOG_INSTALL YES
                    LOG_OUTPUT_ON_FAILURE YES
                    CONFIGURE_COMMAND
                        COMMAND <SOURCE_DIR>/configure
                                --enable-openssl
                                --enable-led
                                --enable-pkcs11
                                --enable-unity
                                --enable-eap-identity
                                --enable-eap-md5
                                --enable-eap-mschapv2
                                --enable-eap-dynamic
                                --enable-eap-tls
                                --enable-eap-ttls
                                --enable-eap-peap
                                --enable-eap-radius
                                --enable-xauth-eap
                                --enable-dhcp
                    BUILD_IN_SOURCE 0 <--- And this !
                    INSTALL_COMMAND make DESTDIR=<INSTALL_DIR> install
)
```
And then I changed to:

```
ExternalProject_Add(strongswan
                    INSTALL_DIR ${CMAKE_BINARY_DIR}/host-install
                    LOG_CONFIGURE YES
                    LOG_BUILD YES
                    LOG_INSTALL YES
                    LOG_OUTPUT_ON_FAILURE YES
                    URL ${CMAKE_CURRENT_SOURCE_DIR}/strongswan-5.9.1 <--- Use URL to copy source code into the build location
                    CONFIGURE_COMMAND
                        COMMAND <SOURCE_DIR>/configure
                                --enable-openssl
                                --enable-led
                                --enable-pkcs11
                                --enable-unity
                                --enable-eap-identity
                                --enable-eap-md5
                                --enable-eap-mschapv2
                                --enable-eap-dynamic
                                --enable-eap-tls
                                --enable-eap-ttls
                                --enable-eap-peap
                                --enable-eap-radius
                                --enable-xauth-eap
                                --enable-dhcp
                    BUILD_IN_SOURCE 1   <----- But build in source so to avoid unpredictable dynamics while producing output files !
                    INSTALL_COMMAND make DESTDIR=<INSTALL_DIR> install
)
```
Thanks again!
